rotkehlchen/externalapis/defillama.py

```diff
@@ -0,0 +1,290 @@
+from rotkehlchen.history.types import HistoricalPrice, HistoricalPriceOracle
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.interfaces import HistoricalPriceOracleInterface
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import Price, Timestamp
+from rotkehlchen.utils.misc import create_timestamp, timestamp_to_date, ts_now
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class DefiLlama(HistoricalPriceOracleInterface):
+
+    def __init__(self) -> None:
+        super().__init__(oracle_name='defillama')
+        self.session = requests.session()
+        self.session.headers.update({'User-Agent': 'rotkehlchen'})
+        self.all_coins_cache: Optional[Dict[str, Dict[str, Any]]] = None
+        self.last_rate_limit = 0
+
+    def _query(
+            self,
+            module: str,
+            subpath: Optional[str] = None,
+            options: Optional[Dict[str, str]] = None,
+    ) -> Dict[str, Any]:
+        """Performs a defillama query
+
+        May raise:
+        - RemoteError if there is a problem querying defillama
+        """
+        if options is None:
+            options = {}
+        url = f'https://coins.llama.fi/{module}/'
+        if subpath:
+            url += subpath
+
+        log.debug(f'Querying defillama: {url}?{urlencode(options)}')
+        try:
+            response = self.session.get(
+                f'{url}?{urlencode(options)}',
+                timeout=DEFAULT_TIMEOUT_TUPLE,
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'DefiLlama API request failed due to {str(e)}') from e
+
+        if response.status_code == HTTPStatus.TOO_MANY_REQUESTS:
+            self.last_rate_limit = ts_now()
+            msg = f'Got rate limited by coingecko querying {url}'
+            log.warning(msg)
+            raise RemoteError(message=msg, error_code=HTTPStatus.TOO_MANY_REQUESTS)
+
+        if response.status_code != 200:
+            msg = (
+                f'DefiLlama API request {response.url} failed with HTTP status '
+                f'code: {response.status_code}'
+            )
+            raise RemoteError(msg)
+
+        try:
+            decoded_json = json.loads(response.text)
+        except json.decoder.JSONDecodeError as e:
+            msg = f'Invalid JSON in Coingecko response. {e}'
+            raise RemoteError(msg) from e
+
+        return decoded_json
+
+    def _get_asset_id(self, asset: AssetWithOracles) -> Optional[str]:
+        """
+        Create the id to be used in DefiLlama.
+        May raise:
+        - UnsupportedAsset
+        """
+        if asset.is_evm_token() is True:
+            asset = asset.resolve_to_evm_token()
+            return f'{str(asset.chain)}:{asset.evm_address}'
+
+        return f'coingecko:{asset.to_coingecko()}'
+
+    def query_current_price(
+            self,
+            from_asset: AssetWithOracles,
```

Comment:
at the begging I thought it could only be evm token but later I tried it and saw that you can query also assets with their coingecko id and it worked for all the assets I tried